gateway-service-livy/src/main/java/org/apache/knox/gateway/livy/LivyHaDispatch.java

```diff
@@ -0,0 +1,156 @@
+           LOG.errorConnectingToServer(outboundRequest.getURI().toString(), e);
+           failoverRequest(outboundRequest, inboundRequest, outboundResponse, inboundResponse, e);
+        }
+     }
+
+    /**
+     * Checks for specific outbound response codes/content to trigger a retry or failover
+     */
+    @Override
+    protected void writeOutboundResponse(HttpUriRequest outboundRequest, HttpServletRequest inboundRequest, HttpServletResponse outboundResponse, HttpResponse inboundResponse) throws IOException {
+       int status = inboundResponse.getStatusLine().getStatusCode();
+       if ( status  == 403 || status == 307) {
+          BufferedHttpEntity entity = new BufferedHttpEntity(inboundResponse.getEntity());
+          inboundResponse.setEntity(entity);
+          ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+          inboundResponse.getEntity().writeTo(outputStream);
+          String body = new String(outputStream.toByteArray(), StandardCharsets.UTF_8);
+          if (body.contains("This is a standby Livy Instance")) {
+             throw new StandbyException();
+          }
+       }
+       super.writeOutboundResponse(outboundRequest, inboundRequest, outboundResponse, inboundResponse);
+    }
+
+    @Override
+    protected void failoverRequest(HttpUriRequest outboundRequest, HttpServletRequest inboundRequest, HttpServletResponse outboundResponse, HttpResponse inboundResponse, Exception exception) throws IOException {
+       LOG.failingOverRequest(outboundRequest.getURI().toString());
+       URI uri;
+       String outboundURIs;
+       AtomicInteger counter = (AtomicInteger) inboundRequest.getAttribute(FAILOVER_COUNTER_ATTRIBUTE);
+       if (counter == null) {
+          counter = new AtomicInteger(0);
+       }
+       inboundRequest.setAttribute(FAILOVER_COUNTER_ATTRIBUTE, counter);
+       outboundURIs = outboundRequest.getURI().toString();
+
+       if (counter.incrementAndGet() <= getMaxFailoverAttempts()) {
+          //null out target url so that rewriters run again
+          inboundRequest.setAttribute(AbstractGatewayFilter.TARGET_REQUEST_URL_ATTRIBUTE_NAME, null);
+
+           uri = getUriFromInbound(inboundRequest, inboundResponse, outboundURIs);
+           ((HttpRequestBase) outboundRequest).setURI(uri);
+          if (getFailoverSleep() > 0) {
+             try {
+                Thread.sleep(getFailoverSleep());
+             } catch (InterruptedException e) {
+                LOG.failoverSleepFailed(getServiceRole(), e);
+             }
+          }
+          executeRequest(outboundRequest, inboundRequest, outboundResponse);
+       } else {
+          LOG.maxFailoverAttemptsReached(getMaxFailoverAttempts(), getServiceRole());
+          if (inboundResponse != null) {
+             writeOutboundResponse(outboundRequest, inboundRequest, outboundResponse, inboundResponse);
+          } else {
+             throw new IOException(exception);
+          }
+       }
+    }
```

Comment:
This whole method is a copy from `DefaultHaDispatch` so don't need to copy it here?

I guess the only differnce is `getUriFromInbound` but that instead could be an overridden `getDispatchUrl` and add the custom logic there?

Reply:
refactored to inherit from DefaultHADispatch